nntrainer/dataset/data_iteration.h

```diff
@@ -0,0 +1,169 @@
+
+  Iteration(const Iteration &rhs) = delete;
+  Iteration &operator=(const Iteration &rhs) = delete;
+  Iteration(Iteration &&rhs) = default;
+  Iteration &operator=(Iteration &&rhs) = default;
+
+  /**
+   * @brief get batch size of iteration
+   *
+   * @return unsigned int batch size
+   */
+  unsigned int batch() { return inputs.front().batch(); }
+
+  /**
+   * @brief Get the Input Reference object
+   *
+   * @return std::vector<Tensor>& input
+   */
+  std::vector<Tensor> &getInputsRef() { return inputs; }
+
+  /**
+   * @brief Get the Input Reference object
+   *
+   * @return const std::vector<Tensor>& input
+   */
+  const std::vector<Tensor> &getInputsRef() const { return inputs; }
+
+  /**
+   * @brief Get the Label Reference object
+   *
+   * @return std::vector<Tensor>&  label
+   */
+  std::vector<Tensor> &getLabelsRef() { return labels; }
+
+  /**
+   * @brief Get the Label Reference object
+   *
+   * @return const std::vector<Tensor>&  label
+   */
+  const std::vector<Tensor> &getLabelsRef() const { return labels; }
+
+  /**
+   * @brief get sample iterator begin()
+   *
+   * @return std::vector<Sample>::iterator
+   */
+  std::vector<Sample>::iterator begin() { return samples.begin(); }
+
+  /**
+   * @brief get sample iterator end
+   *
+   * @return std::vector<Sample>::iterator
+   */
+  std::vector<Sample>::iterator end() { return samples.end(); }
+
+  /**
+   * @brief get sample iterator begin
+   *
+   * @return std::vector<Sample>::const_iterator
+   */
+  std::vector<Sample>::const_iterator begin() const { return samples.end(); }
+
+  /**
+   * @brief get sample iterator end
+   *
+   * @return std::vector<Sample>::const_iterator
+   */
+  std::vector<Sample>::const_iterator end() const { return samples.end(); }
+
+private:
+  std::vector<Tensor> inputs, labels;
+  std::vector<Sample> samples;
+};
+
+/**
+ * @brief Sample class which views the memory for a single sample
+ *
+ */
+class Sample {
+
+public:
+  /**
+   * @brief Construct a new Sample object
+   * @note the batch dimension will be ignored to make a single sample
+   *
+   * @param iter iteration obejcts
+   * @param batch nth batch to create the sample
+   */
+  Sample(const Iteration &iter, unsigned int batch);
+
+  /**
+   * @brief Get the Input Reference object
+   *
+   * @return std::vector<Tensor>& input
+   */
+  std::vector<Tensor> &getInputsRef() { return inputs; }
+
+  /**
+   * @brief Get the Input Reference object
+   *
+   * @return const std::vector<Tensor>& input
+   */
+  const std::vector<Tensor> &getInputsRef() const { return inputs; }
+
+  /**
+   * @brief Get the Label Reference object
+   *
+   * @return std::vector<Tensor>&  label
+   */
+  std::vector<Tensor> &getLabelsRef() { return labels; }
+
+  /**
+   * @brief Get the Label Reference object
+   *
+   * @return const std::vector<Tensor>&  label
+   */
+  const std::vector<Tensor> &getLabelsRef() const { return labels; }
+
+private:
+  std::vector<Tensor> inputs, labels;
```

Comment:
We might in future move out from shared_ptr<float> in Tensor to void * in Tensor.
I recommend to define the ownership of memory at one, and use Tensor * in rest of the places.
It will make the ownership and setting of data clearer.

Reply:
It has the different dimension (is getBatchSlice() of the original tensor) so I guess it has to stay Tensor else no way?